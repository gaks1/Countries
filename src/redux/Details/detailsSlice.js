import {  createSlice } from '@reduxjs/toolkit';

const initialState = {
    countrydetails: [],
    iLoading: false,
    error: null,
};

const detailsSlice = createSlice({
    name: 'details',
    initialState,
    reducers: {},
    extraReducers: {
    },
});

export default detailsSlice.reducer;
