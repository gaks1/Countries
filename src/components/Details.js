import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import { fetchCountry } from '../redux/Details/detailsSlice';
import styles from './Details.module.css';
import DetailsNavbar from './DetailsNavbar';

const Details = () => {
  const { countryName } = useParams();
  const { countrydetails, isLoading } = useSelector((state) => state.details);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchCountry(countryName));
  }, [dispatch, countryName]);

  if (isLoading) {
    return (
      <p className={styles.loading}>Loading...</p>
    );
  }
  return (
    <div className="Details">
      <DetailsNavbar />
      {countrydetails.map(
        (country) => (
          <div key={country.cca3}>
            <div className={styles.divheader}>
              <div key={country.cca3} className={styles.divtitle}>
                <div className={styles.divtitleimg}>
                  <img src={country.flags.png} alt={country.flags.alt} />
                </div>
                <div className={styles.divtitletext}>
                  <h1>{country.name.common}</h1>
                  <p>
                    LAT:
                    {' '}
                    {country.latlng[0]}
                  </p>
                  <p>
                    LONG:
                    {' '}
                    {country.latlng[1]}
                  </p>
                </div>
              </div>
            </div>
            <div className={styles.divseparator}>
              <p className={styles.septext}>COUNTRY DETAILS</p>
            </div>
            <ul className={styles.ul}>
              <li className={styles.liDetails} key={country.cca3}>
                <p className={styles.pdetails}>
                  <span className={styles.pstrong}>Common name</span>
                  <span className={styles.pproperty}>{country.name.common}</span>
                </p>
                <p className={styles.pdetails}>
                  <span className={styles.pstrong}>Official name</span>
                  <span className={styles.pproperty}>{country.name.official}</span>
                </p>
                <p className={styles.pdetails}>
                  <span className={styles.pstrong}>Capital</span>
                  <span className={styles.pproperty}>
                    {Array.isArray(country.capital)
                      ? country.capital.join(',  ')
                      : country.capital}
                  </span>
                </p>
                <p className={styles.pdetails}>
                  <span className={styles.pstrong}>Country region</span>
                  <span className={styles.pproperty}>{country.region}</span>
                </p>
                <p className={styles.pdetails}>
                  <span className={styles.pstrong}>Country-subregion</span>
                  <span className={styles.pproperty}>{country.subregion}</span>
                </p>
                <p className={styles.pdetails}>
                  <span className={styles.pstrong}>Country-population</span>
                  <span className={styles.pproperty}>{country.population}</span>
                </p>
                <p className={styles.pdetails}>
                  <span className={styles.pstrong}>Country-area</span>
                  <span className={styles.pproperty}>{country.area}</span>
                </p>
                <p className={styles.pdetails}>
                  <span className={styles.pstrong}>Continent</span>
                  <span className={styles.pproperty}>{country.continents}</span>
                </p>
                <p className={styles.pdetails}>
                  <span className={styles.pstrong}>Timezones</span>
                  <span className={styles.pproperty}>{country.timezones.join(',  ')}</span>
                </p>
                <p className={styles.pdetails}>
                  <span className={styles.pstrong}>Start of the week</span>
                  <span className={styles.pproperty}>{country.startOfWeek}</span>
                </p>
                <p className={styles.pdetails}>
                  <span className={styles.pstrong}>Street view</span>
                  <a href={country.maps.googleMaps} className={styles.pproperty}>google maps</a>
                </p>
                <p className={styles.pdetails}>
                  <span className={styles.pstrong}>Language:</span>
                  {
                    country.languages && (
                      <ul className={styles.languageContainer}>
                        {Object.entries(country.languages).map(([code, name]) => (
                          <li key={code} className={styles.languageItem}>{name}</li>
                        ))}
                      </ul>
                    )
                  }
                </p>

                <p className={styles.pdetails}>
                  <span className={styles.pstrong}>Currency:</span>
                  {
                    country.currencies && (
                      <ul className={styles.languageContainer}>
                        {Object.entries(country.currencies).map(([code, currency]) => (
                          <li key={code} className={styles.languageItem}>
                            {currency.name}
                            {' '}
                            (
                            {currency.symbol}
                            )
                          </li>
                        ))}
                      </ul>
                    )
                  }
                </p>

              </li>
            </ul>
          </div>
        ),
      )}

    </div>
  );
};

export default Details;
