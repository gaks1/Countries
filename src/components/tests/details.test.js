import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import {
  Routes, Route, MemoryRouter,
} from 'react-router-dom';

import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Details from '../Details';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('Details', () => {
  const initialState = {
    details: {
      countrydetails: [
        {
          cca3: 'AFG',
          name: {
            common: 'Afghanistan',
            official: 'Islamic Republic of Afghanistan',
          },
          flags: {
            png: 'https://restcountries.com/data/afg.svg',
            alt: 'flag',
          },
          latlng: [20, 20],
          capital: ['Kabul'],
          region: 'Asia',
          subregion: 'Southern Asia',
          population: 32225560,
          area: 652230,
          timezones: ['UTC+04:30'],
          startOfWeek: 'Saturday',
          maps: {
            googleMaps: 'https://www.google.com/maps/place/Afghanistan',
          },
          languages: { ps: 'Pashto', uz: 'Uzbek', tk: 'Turkmen' },
          currencies: {
            AFN: {
              name: 'Afghan afghani',
              symbol: 'Ø‹',
            },
          },
        },
      ],
      isLoading: false,
    },
  };

  const store = mockStore(initialState);
  const countryName = 'Afghanistan';

  test('renders without crashing', () => {
    const { getByText } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={[`/${countryName}`]}>
          <Routes>
            <Route path="/:countryName" element={<Details />} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    );
    expect(getByText('Saturday')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <MemoryRouter initialEntries={[`/${countryName}`]}>
          <Routes>
            <Route path="/:countryName" element={<Details />} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
