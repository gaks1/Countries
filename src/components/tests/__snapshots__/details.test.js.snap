// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details matches snapshot 1`] = `
<div
  className="Details"
>
  <nav
    className="navbar"
  >
    <div
      className="divbutton"
    >
      <button
        className="button"
        onClick={[Function]}
        type="button"
      >
        <span
          className="material-symbols-outlined"
        >
          arrow_back_ios
        </span>
      </button>
    </div>
    <div>
      <h1
        className="h1title"
      >
        DETAILS
      </h1>
    </div>
  </nav>
  <div>
    <div
      className="divheader"
    >
      <div
        className="divtitle"
      >
        <div
          className="divtitleimg"
        >
          <img
            alt="flag"
            src="https://restcountries.com/data/afg.svg"
          />
        </div>
        <div
          className="divtitletext"
        >
          <h1>
            Afghanistan
          </h1>
          <p>
            LAT:
             
            20
          </p>
          <p>
            LONG:
             
            20
          </p>
        </div>
      </div>
    </div>
    <div
      className="divseparator"
    >
      <p
        className="septext"
      >
        COUNTRY DETAILS
      </p>
    </div>
    <ul
      className="ul"
    >
      <li
        className="liDetails"
      >
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Common name
          </span>
          <span
            className="pproperty"
          >
            Afghanistan
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Official name
          </span>
          <span
            className="pproperty"
          >
            Islamic Republic of Afghanistan
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Capital
          </span>
          <span
            className="pproperty"
          >
            Kabul
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Country region
          </span>
          <span
            className="pproperty"
          >
            Asia
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Country-subregion
          </span>
          <span
            className="pproperty"
          >
            Southern Asia
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Country-population
          </span>
          <span
            className="pproperty"
          >
            32225560
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Country-area
          </span>
          <span
            className="pproperty"
          >
            652230
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Continent
          </span>
          <span
            className="pproperty"
          />
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Timezones
          </span>
          <span
            className="pproperty"
          >
            UTC+04:30
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Start of the week
          </span>
          <span
            className="pproperty"
          >
            Saturday
          </span>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Street view
          </span>
          <a
            className="pproperty"
            href="https://www.google.com/maps/place/Afghanistan"
          >
            google maps
          </a>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Language:
          </span>
          <ul
            className="languageContainer"
          >
            <li
              className="languageItem"
            >
              Pashto
            </li>
            <li
              className="languageItem"
            >
              Uzbek
            </li>
            <li
              className="languageItem"
            >
              Turkmen
            </li>
          </ul>
        </p>
        <p
          className="pdetails"
        >
          <span
            className="pstrong"
          >
            Currency:
          </span>
          <ul
            className="languageContainer"
          >
            <li
              className="languageItem"
            >
              Afghan afghani
               
              (
              Ø‹
              )
            </li>
          </ul>
        </p>
      </li>
    </ul>
  </div>
</div>
`;
